diff -ruN linux/drivers/platform/mips/yeeloong_laptop.c linux-fixed/drivers/platform/mips/yeeloong_laptop.c
--- linux/drivers/platform/mips/yeeloong_laptop.c	2014-01-01 19:21:06.356567382 +0800
+++ linux-fixed/drivers/platform/mips/yeeloong_laptop.c	2014-01-01 19:21:33.672607421 +0800
@@ -168,7 +168,7 @@
 
 static int get_bat_temp(void)
 {
-	return get_bat_l(TEMPERATURE) * 1000;
+	return get_bat_l(TEMPERATURE) * 10;
 }
 
 static int get_bat_current(void)
