diff -rupN linux-3.15-orig/Makefile linux-3.15-new/Makefile
--- linux-3.15-orig/Makefile	2014-07-03 05:54:19.662112107 +0800
+++ linux-3.15-new/Makefile	2014-07-03 05:55:13.542027915 +0800
@@ -396,6 +396,9 @@ KBUILD_CFLAGS   := -Wall -Wundef -Wstric
 		   -fno-strict-aliasing -fno-common \
 		   -Werror-implicit-function-declaration \
 		   -Wno-format-security \
+		   -ftree-vectorize -pipe \
+		   -floop-interchange -ftree-loop-distribution \
+		   -floop-strip-mine -floop-block \
 		   $(call cc-option,-fno-delete-null-pointer-checks,)
 KBUILD_AFLAGS_KERNEL :=
 KBUILD_CFLAGS_KERNEL :=
